### PR TITLE
Iss16

### PR DESCRIPTION
**Pull request by qxp134**
update iss16&iss17

### CODE REVIEW

File: 16.cpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+#include <algorithm>
+#include <iostream>
+#include <vector>
+
+class Solution {
+ public:
+  int threeSumClosest(std::vector<int>& nums, int target) {
+    sort(nums.begin(), nums.end());
+    int last_num = 999;
```
Comment (chenleexyz) [inline 16.cpp:10]: ??? NOTE `last_num` is never updated.

File: 16.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
+// https://leetcode-cn.com/problems/3sum-closest/
+#include <limits.h>
```
Comment (chenleexyz) [inline 16.cpp]: useless? BTW, use `<climits>` instead. `<limits>` is better.

File: 16.cpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+class Solution {
+ public:
+  int threeSumClosest(std::vector<int>& nums, int target) {
+    sort(nums.begin(), nums.end());
+    int last_num = 999;
+
+    int len = nums.size();
+    int ans = nums[0] + nums[len - 1] + nums[len - 2];
+    int min_diff = abs(target - ans);
```
Comment (ilichen) [inline 16.cpp:16]: I sugest, rewrite as 
`int min_diff = std::numeric_limits<int>::max();`
this way, the code is more clean.

File: 16.cpp [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    for (int i = 0; i < len; ++i) {
+      if (nums[i] == last_num) {
+        continue;
+      }
```
Comment (chenleexyz) [inline 16.cpp]: ???

File: 16.cpp [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+#include <limits.h>
+
+#include <algorithm>
+#include <iostream>
+#include <vector>
+
+class Solution {
+ public:
+  int threeSumClosest(std::vector<int>& nums, int target) {
+    sort(nums.begin(), nums.end());
```
Comment (chenleexyz) [inline 16.cpp]: `std::sort`

File: 16.cpp [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        if (diff > 0) {
+          --k;
+        } else if (diff < 0) {
+          ++j;
+        } else {
+          return target;
+        }
+        if (abs(diff) < min_diff) {
+          ans = target + diff;
+          min_diff = abs(diff);  // frist commit I forget abs
```
Comment (chenleexyz) [inline 16.cpp]: `std::abs`

File: 16.cpp (hunk 3 above)
Comment (chenleexyz) [inline 16.cpp]: `ans` -> `best_ans` / `closest`

File: 16.cpp [hunk 7]
```diff
@@ -0,0 +1,41 @@
+// https://leetcode-cn.com/problems/3sum-closest/
+#include <limits.h>
+
+#include <algorithm>
+#include <iostream>
+#include <vector>
+
+class Solution {
+ public:
+  int threeSumClosest(std::vector<int>& nums, int target) {
```
Comment (chenleexyz) [inline 16.cpp]: `const std::vector<int>&`

File: 17.cpp [hunk 8]
```diff
@@ -0,0 +1,50 @@
+// https://leetcode-cn.com/problems/letter-combinations-of-a-phone-number/
+#include <iostream>
+#include <string>
+#include <vector>
+std::vector<std::string> res;
```
Comment (chenleexyz) [inline 17.cpp]: Don't use global vars.

File: 17.cpp [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+std::vector<std::string> recursion(std::string digits) {
+  if (digits.empty()) return res;
+  std::string::iterator it = digits.begin();
+  int count = 3;
+  if (*it != '1') {
+    if (*it == '7' || *it == '9') {  // forget num '7'
+      count = 4;
+    }
```
Comment (chenleexyz) [inline 17.cpp]: I suggest create a table to handle this
```
std::string number_to_chars[10] = { "", "", "abc", ....};
```

**Comment by chenleexyz** [comment]
Prefer a meaningful var name, rather than `result`, `res` if you can.

File: 17.cpp [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    for (int i = 0; i < count; ++i) {
+      temp_arr[i] = res;
+      if (temp_arr[i].empty()) {
+        std::string temp_str(1, start_char);
+        temp_arr[i].push_back(temp_str);
+      } else {
+        for (auto iter = temp_arr[i].begin(); iter != temp_arr[i].end();
+             ++iter) {
+          iter->push_back(start_char);
+        }
```
Comment (chenleexyz) [inline 17.cpp]:
```C++
for (std::string& str : temp_arr[i]) str.push_back(start_char);
```

File: 16.cpp (hunk 2 above)
Comment (qxp134) [reply 16.cpp]: forget delete

File: 16.cpp (hunk 1 above)
Comment (qxp134) [reply 16.cpp:10]: I forget use this variable

File: 16.cpp (hunk 4 above)
Comment (qxp134) [reply 16.cpp]: if one number equal with last number, skip it

File: 16.cpp (hunk 3 above)
Comment (qxp134) [reply 16.cpp]: ok

File: 17.cpp (hunk 8 above)
Comment (qxp134) [reply 17.cpp]: I want save memory and get a higher score

**Review by qxp134 — COMMENTED** [review]
fix those problems

File: 16.cpp (hunk 7 above)
Comment (chenleexyz) [reply 16.cpp]: Oops, I'm wrong, `std::sort` will modify nums.

**Comment by chenleexyz** [comment]
LGTM

**Pull request merged**